File: tests/100_basic_work.cpp
#include <nan.h>
#include <ncd.hpp>

#include <unistd.h>

namespace {

void
testWorkQueue(Nan::FunctionCallbackInfo<v8::Value> const& args) {
  Nan::HandleScope scope;
  unsigned items = args[0]->Uint32Value();
  unsigned delay = args[1]->Uint32Value();
  
  for (unsigned i = 0; i < items; ++i) {
    dispatch(ncd::defaultWorkQueue(), [=](){
      if (delay != 0) {
        usleep(delay * 1000);
      }
    }, args[2].As<v8::Function>());
  }
}

void
testStringResult(Nan::FunctionCallbackInfo<v8::Value> const& args) {
  dispatch(ncd::defaultWorkQueue(),
      [=](){ return "This is fine."; },
      args[0].As<v8::Function>());
}

void
testDoubleResult(Nan::FunctionCallbackInfo<v8::Value> const& args) {
  dispatch(ncd::defaultWorkQueue(),
      [=](){ return 0.5; },
      args[0].As<v8::Function>());
}

void
testStringOrError(Nan::FunctionCallbackInfo<v8::Value> const& args) {
  bool succeed = args[0]->ToBoolean()->Value();
  dispatch(ncd::defaultWorkQueue(), [=](ncd::AsyncError ** error) {
    if (succeed) {
      return "This is fine.";
    } else {
      *error = new ncd::AsyncError("Kaputt.");
      return (char const*)nullptr;
    }
  }, args[1].As<v8::Function>());
}

void
init(v8::Local<v8::Object> exports) {
  exports->Set(ncd::v8str("testWorkQueue"), ncd::function(testWorkQueue));
  exports->Set(ncd::v8str("testStringResult"), ncd::function(testStringResult));
  exports->Set(ncd::v8str("testDoubleResult"), ncd::function(testDoubleResult));
  exports->Set(ncd::v8str("testStringOrError"), ncd::function(testStringOrError));
}

}  // end of anonymous namespace

NCD_NODE_ADDON(init)

File: tests/100_basic_work.js
const
  basic_work = require('./build/Release/100_basic_work'),
  tap = require('tap')

tap.test("work queue: many items", function (t) {
  var doneCount = 0,
    testItems = 100000,
    delay = 0
  t.plan(testItems + 1)
  basic_work.testWorkQueue(testItems, delay, () => { 
    t.pass("done callback called after work")
    if (++doneCount == testItems) {
      setTimeout(() => {
        t.ok(doneCount == testItems)
        t.end()
      }, 3 * delay + 100)
    }
  })
})

tap.test(function (t) {
  var doneCount = 0,
    testItems = 1000,
    delay = 1
  t.plan(testItems + 1)
  basic_work.testWorkQueue(testItems, delay, () => { 
    t.pass("done callback called after work")
    if (++doneCount == testItems) {
      setTimeout(() => {
        t.ok(doneCount == testItems)
        t.end()
      }, 3 * delay + 100)
    }
  })
})

tap.test((t) => {
  t.plan(1)
  basic_work.testStringResult((result) => {
    t.equal(result, "This is fine.")
    t.end()
  })
})

tap.test((t) => {
  t.plan(1)
  basic_work.testDoubleResult((result) => {
    t.equal(result, 0.5)
    t.end()
  })
})

tap.test((t) => {
  var count = 50000, done = 0
  t.plan(2 * count)
  function handler(success) {
    if (success) {
      return (error, result) => {
        t.equal(error, null)
        t.equal(result, "This is fine.")
        if (++done == count) { t.end() }
      } 
    } else {
      return (error, result) => {
        t.equal(error.message, "Kaputt.")
        t.equal(result, null)
        if (++done == count) { t.end() }
      }
    }
  }
  for (var i = 0; i < count; ++i) {
    var succeed = Math.random() < 0.5
    basic_work.testStringOrError(succeed, handler(succeed))
  }
})
